### PR TITLE
Add insert_materialized_view_atomic to control pushing to MV behaviour

### PR DESCRIPTION
**Pull request by azat**
Changelog category (leave one):
- Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Add insert_materialized_view_atomic to control pushing to MV behaviour (by default true, if set to false, will require less memory if there are multiple underlying MVs)

Detailed description / Documentation draft:
By default push to MV is atomic, that said that if push to one of the MV
fails then none of MVs will be updated, while this requires memory for
pendings blocks to all MVs and this can be pretty huge (depends on the
number of MVs)

So add a setting to control this behaviour, and prefer atomicity by
default.

### CODE REVIEW

**Comment by azat** [comment]
Fix `readability-qualified-auto`

*Previous version of the patch is [here](https://github.com/ClickHouse/ClickHouse/commit/e12acd881a78f8d30926caba017bac5e8a370355), attached for CI report*

**Comment by azat** [comment]
>Functional stateless tests (thread) — fail: 20, passed: 125, skipped: 1

<details>

```
==================
WARNING: ThreadSanitizer: data race (pid=209)
  Write of size 4 at 0x000015329e08 by thread T97 (mutexes: write M2950919, write M84014834532131392, write M84577784487578640):
    #0 cpuid_get_raw_data /build/obj-x86_64-linux-gnu/../contrib/libcpuid/libcpuid/cpuid_main.c (clickhouse+0x10e6368a)
    #1 getNumberOfPhysicalCPUCores() /build/obj-x86_64-linux-gnu/../src/Common/getNumberOfPhysicalCPUCores.cpp:18:14 (clickhouse+0x96b1060)
    #2 DB::InterpreterSystemQuery::restartReplicas(DB::Context&) /build/obj-x86_64-linux-gnu/../src/Interpreters/InterpreterSystemQuery.cpp:388:37 (clickhouse+0x1039b3c3)
    #3 DB::InterpreterSystemQuery::execute() /build/obj-x86_64-linux-gnu/../src/Interpreters/InterpreterSystemQuery.cpp:290:13 (clickhouse+0x103969b0)
    #4 DB::executeQueryImpl(char const*, char const*, DB::Context&, bool, DB::QueryProcessingStage::Enum, bool, DB::ReadBuffer*, bool) /build/obj-x86_64-linux-gnu/../src/Interpreters/executeQuery.cpp:343:32 (clickhouse+0x1049edcc)
    #5 DB::executeQuery(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, DB::Context&, bool, DB::QueryProcessingStage::Enum, bool, bool) /build/obj-x86_64-linux-gnu/../src/Interpreters/executeQuery.cpp:588:30 (clickhouse+0x1049ddbe)
    #6 DB::TCPHandler::runImpl() /build/obj-x86_64-linux-gnu/../programs/server/TCPHandler.cpp:252:24 (clickhouse+0x971071c)
    #7 DB::TCPHandler::run() /build/obj-x86_64-linux-gnu/../programs/server/TCPHandler.cpp:1242:9 (clickhouse+0x971d2a7)
    #8 Poco::Net::TCPServerConnection::start() /build/obj-x86_64-linux-gnu/../contrib/poco/Net/src/TCPServerConnection.cpp:43:3 (clickhouse+0x136e97b2)
    #9 Poco::Net::TCPServerDispatcher::run() /build/obj-x86_64-linux-gnu/../contrib/poco/Net/src/TCPServerDispatcher.cpp:114:20 (clickhouse+0x136e9ff6)
    #10 Poco::PooledThread::run() /build/obj-x86_64-linux-gnu/../contrib/poco/Foundation/src/ThreadPool.cpp:199:14 (clickhouse+0x13852e66)
    #11 Poco::(anonymous namespace)::RunnableHolder::run() /build/obj-x86_64-linux-gnu/../contrib/poco/Foundation/src/Thread.cpp:55:11 (clickhouse+0x1385159f)
    #12 Poco::ThreadImpl::runnableEntry(void*) /build/obj-x86_64-linux-gnu/../contrib/poco/Foundation/src/Thread_POSIX.cpp:345:27 (clickhouse+0x1384fe27)
    #13 <null> <null> (clickhouse+0x957970c)

  Previous write of size 4 at 0x000015329e08 by thread T281 (mutexes: write M341845913198728112, write M82325984669510080, write M83170409602800592):
    #0 cpuid_get_raw_data /build/obj-x86_64-linux-gnu/../contrib/libcpuid/libcpuid/cpuid_main.c (clickhouse+0x10e6368a)
    #1 getNumberOfPhysicalCPUCores() /build/obj-x86_64-linux-gnu/../src/Common/getNumberOfPhysicalCPUCores.cpp:18:14 (clickhouse+0x96b1060)
    #2 DB::InterpreterSystemQuery::restartReplicas(DB::Context&) /build/obj-x86_64-linux-gnu/../src/Interpreters/InterpreterSystemQuery.cpp:388:37 (clickhouse+0x1039b3c3)
    #3 DB::InterpreterSystemQuery::execute() /build/obj-x86_64-linux-gnu/../src/Interpreters/InterpreterSystemQuery.cpp:290:13 (clickhouse+0x103969b0)
    #4 DB::executeQueryImpl(char const*, char const*, DB::Context&, bool, DB::QueryProcessingStage::Enum, bool, DB::ReadBuffer*, bool) /build/obj-x86_64-linux-gnu/../src/Interpreters/executeQuery.cpp:343:32 (clickhouse+0x1049edcc)
    #5 DB::executeQuery(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, DB::Context&, bool, DB::QueryProcessingStage::Enum, bool, bool) /build/obj-x86_64-linux-gnu/../src/Interpreters/executeQuery.cpp:588:30 (clickhouse+0x1049ddbe)
    #6 DB::TCPHandler::runImpl() /build/obj-x86_64-linux-gnu/../programs/server/TCPHandler.cpp:252:24 (clickhouse+0x971071c)
    #7 DB::TCPHandler::run() /build/obj-x86_64-linux-gnu/../programs/server/TCPHandler.cpp:1242:9 (clickhouse+0x971d2a7)
    #8 Poco::Net::TCPServerConnection::start() /build/obj-x86_64-linux-gnu/../contrib/poco/Net/src/TCPServerConnection.cpp:43:3 (clickhouse+0x136e97b2)
    #9 Poco::Net::TCPServerDispatcher::run() /build/obj-x86_64-linux-gnu/../contrib/poco/Net/src/TCPServerDispatcher.cpp:114:20 (clickhouse+0x136e9ff6)
    #10 Poco::PooledThread::run() /build/obj-x86_64-linux-gnu/../contrib/poco/Foundation/src/ThreadPool.cpp:199:14 (clickhouse+0x13852e66)
    #11 Poco::(anonymous namespace)::RunnableHolder::run() /build/obj-x86_64-linux-gnu/../contrib/poco/Foundation/src/Thread.cpp:55:11 (clickhouse+0x1385159f)
    #12 Poco::ThreadImpl::runnableEntry(void*) /build/obj-x86_64-linux-gnu/../contrib/poco/Foundation/src/Thread_POSIX.cpp:345:27 (clickhouse+0x1384fe27)
    #13 <null> <null> (clickhouse+0x957970c)

  Location is global '??' at 0x000000000000 (clickhouse+0x000015329e08)

  Mutex M2950919 (0x7b0c0031d940) created at:
    #0 pthread_mutex_lock <null> (clickhouse+0x9596b76)
    #1 std::__1::__libcpp_mutex_lock(pthread_mutex_t*) /build/obj-x86_64-linux-gnu/../contrib/libcxx/include/__threading_support:322:10 (clickhouse+0x14707699)
    #2 std::__1::mutex::lock() /build/obj-x86_64-linux-gnu/../contrib/libcxx/src/mutex.cpp:33:14 (clickhouse+0x14707699)
    #3 std::__1::unique_lock<std::__1::mutex>::unique_lock(std::__1::mutex&) /build/obj-x86_64-linux-gnu/../contrib/libcxx/include/__mutex_base:119:61 (clickhouse+0xfda0dcc)
    #4 DB::DDLGuard::DDLGuard(std::__1::map<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, DB::DDLGuard::Entry, std::__1::less<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, std::__1::allocator<std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const, DB::DDLGuard::Entry> > >&, std::__1::unique_lock<std::__1::mutex>, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) /build/obj-x86_64-linux-gnu/../src/Interpreters/DatabaseCatalog.cpp:709:18 (clickhouse+0xfda0dcc)
    #5 std::__1::__unique_if<DB::DDLGuard>::__unique_single std::__1::make_unique<DB::DDLGuard, std::__1::map<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, DB::DDLGuard::Entry, std::__1::less<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, std::__1::allocator<std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const, DB::DDLGuard::Entry> > >&, std::__1::unique_lock<std::__1::mutex>, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&>(std::__1::map<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> >, DB::DDLGuard::Entry, std::__1::less<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > >, std::__1::allocator<std::__1::pair<std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const, DB::DDLGuard::Entry> > >&, std::__1::unique_lock<std::__1::mutex>&&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) /build/obj-x86_64-linux-gnu/../contrib/libcxx/include/memory:3028:32 (clickhouse+0xfd9c465)
    #6 DB::DatabaseCatalog::getDDLGuard(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) /build/obj-x86_64-linux-gnu/../src/Interpreters/DatabaseCatalog.cpp:462:12 (clickhouse+0xfd9c465)
    #7 DB::InterpreterRenameQuery::execute() /build/obj-x86_64-linux-gnu/../src/Interpreters/InterpreterRenameQuery.cpp:76:47 (clickhouse+0x100eb90f)
    #8 DB::executeQueryImpl(char const*, char const*, DB::Context&, bool, DB::QueryProcessingStage::Enum, bool, DB::ReadBuffer*, bool) /build/obj-x86_64-linux-gnu/../src/Interpreters/executeQuery.cpp:343:32 (clickhouse+0x1049edcc)
    #9 DB::executeQuery(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, DB::Context&, bool, DB::QueryProcessingStage::Enum, bool, bool) /build/obj-x86_64-linux-gnu/../src/Interpreters/executeQuery.cpp:588:30 (clickhouse+0x1049ddbe)
    #10 DB::TCPHandler::runImpl() /build/obj-x86_64-linux-gnu/../programs/server/TCPHandler.cpp:252:24 (clickhouse+0x971071c)
    #11 DB::TCPHandler::run() /build/obj-x86_64-linux-gnu/../programs/server/TCPHandler.cpp:1242:9 (clickhouse+0x971d2a7)
    #12 Poco::Net::TCPServerConnection::start() /build/obj-x86_64-linux-gnu/../contrib/poco/Net/src/TCPServerConnection.cpp:43:3 (clickhouse+0x136e97b2)
    #13 Poco::Net::TCPServerDispatcher::run() /build/obj-x86_64-linux-gnu/../contrib/poco/Net/src/TCPServerDispatcher.cpp:114:20 (clickhouse+0x136e9ff6)
    #14 Poco::PooledThread::run() /build/obj-x86_64-linux-gnu/../contrib/poco/Foundation/src/ThreadPool.cpp:199:14 (clickhouse+0x13852e66)
    #15 Poco::(anonymous namespace)::RunnableHolder::run() /build/obj-x86_64-linux-gnu/../contrib/poco/Foundation/src/Thread.cpp:55:11 (clickhouse+0x1385159f)
    #16 Poco::ThreadImpl::runnableEntry(void*) /build/obj-x86_64-linux-gnu/../contrib/poco/Foundation/src/Thread_POSIX.cpp:345:27 (clickhouse+0x1384fe27)
    #17 <null> <null> (clickhouse+0x957970c)

  Mutex M84014834532131392 is already destroyed.

  Mutex M84577784487578640 is already destroyed.

  Mutex M341845913198728112 is already destroyed.

  Mutex M82325984669510080 is already destroyed.

  Mutex M83170409602800592 is already destroyed.

  Thread T97 'TCPHandler' (tid=977, running) created by thread T39 at:
    #0 pthread_create <null> (clickhouse+0x95797ab)
    #1 Poco::ThreadImpl::startImpl(Poco::SharedPtr<Poco::Runnable, Poco::ReferenceCounter, Poco::ReleasePolicy<Poco::Runnable> >) /build/obj-x86_64-linux-gnu/../contrib/poco/Foundation/src/Thread_POSIX.cpp:202:6 (clickhouse+0x1384f757)
    #2 Poco::Thread::start(Poco::Runnable&) /build/obj-x86_64-linux-gnu/../contrib/poco/Foundation/src/Thread.cpp:128:2 (clickhouse+0x13850f9c)
    #3 Poco::PooledThread::start() /build/obj-x86_64-linux-gnu/../contrib/poco/Foundation/src/ThreadPool.cpp:85:10 (clickhouse+0x13854baf)
    #4 Poco::ThreadPool::getThread() /build/obj-x86_64-linux-gnu/../contrib/poco/Foundation/src/ThreadPool.cpp:461:14 (clickhouse+0x13854baf)
    #5 Poco::ThreadPool::startWithPriority(Poco::Thread::Priority, Poco::Runnable&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) /build/obj-x86_64-linux-gnu/../contrib/poco/Foundation/src/ThreadPool.cpp:365:2 (clickhouse+0x13854f97)
    #6 Poco::Net::TCPServerDispatcher::enqueue(Poco::Net::StreamSocket const&) /build/obj-x86_64-linux-gnu/../contrib/poco/Net/src/TCPServerDispatcher.cpp:144:17 (clickhouse+0x136ea546)
    #7 Poco::Net::TCPServer::run() /build/obj-x86_64-linux-gnu/../contrib/poco/Net/src/TCPServer.cpp:148:21 (clickhouse+0x136e90f7)
    #8 Poco::(anonymous namespace)::RunnableHolder::run() /build/obj-x86_64-linux-gnu/../contrib/poco/Foundation/src/Thread.cpp:55:11 (clickhouse+0x1385159f)
    #9 Poco::ThreadImpl::runnableEntry(void*) /build/obj-x86_64-linux-gnu/../contrib/poco/Foundation/src/Thread_POSIX.cpp:345:27 (clickhouse+0x1384fe27)
    #10 <null> <null> (clickhouse+0x957970c)

  Thread T281 'TCPHandler' (tid=2019, running) created by thread T39 at:
    #0 pthread_create <null> (clickhouse+0x95797ab)
    #1 Poco::ThreadImpl::startImpl(Poco::SharedPtr<Poco::Runnable, Poco::ReferenceCounter, Poco::ReleasePolicy<Poco::Runnable> >) /build/obj-x86_64-linux-gnu/../contrib/poco/Foundation/src/Thread_POSIX.cpp:202:6 (clickhouse+0x1384f757)
    #2 Poco::Thread::start(Poco::Runnable&) /build/obj-x86_64-linux-gnu/../contrib/poco/Foundation/src/Thread.cpp:128:2 (clickhouse+0x13850f9c)
    #3 Poco::PooledThread::start() /build/obj-x86_64-linux-gnu/../contrib/poco/Foundation/src/ThreadPool.cpp:85:10 (clickhouse+0x13854baf)
    #4 Poco::ThreadPool::getThread() /build/obj-x86_64-linux-gnu/../contrib/poco/Foundation/src/ThreadPool.cpp:461:14 (clickhouse+0x13854baf)
    #5 Poco::ThreadPool::startWithPriority(Poco::Thread::Priority, Poco::Runnable&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) /build/obj-x86_64-linux-gnu/../contrib/poco/Foundation/src/ThreadPool.cpp:365:2 (clickhouse+0x13854f97)
    #6 Poco::Net::TCPServerDispatcher::enqueue(Poco::Net::StreamSocket const&) /build/obj-x86_64-linux-gnu/../contrib/poco/Net/src/TCPServerDispatcher.cpp:144:17 (clickhouse+0x136ea546)
    #7 Poco::Net::TCPServer::run() /build/obj-x86_64-linux-gnu/../contrib/poco/Net/src/TCPServer.cpp:148:21 (clickhouse+0x136e90f7)
    #8 Poco::(anonymous namespace)::RunnableHolder::run() /build/obj-x86_64-linux-gnu/../contrib/poco/Foundation/src/Thread.cpp:55:11 (clickhouse+0x1385159f)
    #9 Poco::ThreadImpl::runnableEntry(void*) /build/obj-x86_64-linux-gnu/../contrib/poco/Foundation/src/Thread_POSIX.cpp:345:27 (clickhouse+0x1384fe27)
    #10 <null> <null> (clickhouse+0x957970c)

SUMMARY: ThreadSanitizer: data race /build/obj-x86_64-linux-gnu/../contrib/libcpuid/libcpuid/cpuid_main.c in cpuid_get_raw_data
==================
```

</details>

Looks like related to https://github.com/ClickHouse/ClickHouse/pull/10396#issuecomment-626213350, i.e. not related to this PR

File: src/DataStreams/PushingToViewsBlockOutputStream.cpp [hunk 1]
```diff
@@ -266,6 +266,9 @@ void PushingToViewsBlockOutputStream::process(const Block & block, size_t view_n
         }
 
         in->readSuffix();
+
+        if (!settings.insert_materialized_view_atomic)
+            view.out->writeSuffix();
```
Comment (azat) [inline src/DataStreams/PushingToViewsBlockOutputStream.cpp:271]: Egh, this is wrong

File: src/DataStreams/PushingToViewsBlockOutputStream.cpp [hunk 2]
```diff
@@ -190,16 +196,19 @@ void PushingToViewsBlockOutputStream::writeSuffix()
...
         {
-            ex.addMessage("while write prefix to view " + view.table_id.getNameForLogs());
-            throw;
+            try
+            {
+                view.out->writeSuffix();
+            }
+            catch (Exception & ex)
+            {
+                ex.addMessage("while write prefix to view " + view.table_id.getNameForLogs());
```
[the remark this answers is not in the document]
Comment (filimonov) [reply src/DataStreams/PushingToViewsBlockOutputStream.cpp:209]: But if we write suffix to MV `a`, `b` and get an exception on `c` we can't rollback `a` and `b`. Is it an amoticity?

**Review by filimonov — COMMENTED** [review]
Not sure if i understand that proberly, and i have doibts if it should be called atomicity...

File: src/DataStreams/PushingToViewsBlockOutputStream.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (azat) [reply src/DataStreams/PushingToViewsBlockOutputStream.cpp:209]: No

**Comment by azat** [comment]
>Not sure if i understand that proberly

Yes, this is unclear and icky.

> i have doibts if it should be called atomicity...

Actually more or less the same (by some reason it uses a little bit more memory for now, but this is a separate issue) this can be archived by introducing separate `min_insert_block_size_rows`/`min_insert_block_size_bytes` for pushing to materialized views (and setting them to `1` will drop extra memory usage, since this will just proxy the data)

**Comment by azat** [comment]
>(by some reason it uses a little bit more memory for now, but this is a separate issue)

That was because some engines bufffer the write

**Comment by azat** [comment]
>this can be archived by introducing separate min_insert_block_size_rows/min_insert_block_size_bytes for pushing to materialized views (and setting them to 1 will drop extra memory usage, since this will just proxy the data)

Or by adding something like `insert_materialized_view_no_squash`, but this is less flexible

**Pull request closed without merging**